icefoganalytics/internal-data-portal
# Dataset Visualize Manage & Read Page

Relates to:
- [0001, Data Modeling, 2024-01-05](https://docs.google.com/document/d/1Mdbl_Cy0BgKiq0DGdvOKUqJn-FdgbF0LSVykwf__FY8/edit?pli=1)
- [0002, UI Modeling, 2024-01-08](https://docs.google.com/document/d/1tP63Ob4XPBGlASPHVf1GnNb3hlIkybmWBqJDA61u3J0/edit)
- Partially completes https://github.com/icefoganalytics/internal-data-portal/issues/54

# Context

Pre-implementation re-working

- [x] Drop SearchFieldExclusion model and use boolean on DatasetField instead.
- [x] Upgrade add API link button to handle JMESPath expressions parsing of data returned from api.
    Or use some helful UI or something.
- [x] ~Re-imagation "limit data view on search" as "show/hide" search.~ Change to enable/disable search customization
- [x] ~drop "row limit" feature and use pagination of all data instead.~ Change to limiting search results when searching


## Dataset Visualize Manage Page - /dataset/:slug/visualize/manage

### Front-end routes

Various tabs at the same level as this page.

### Back-end routes

* /api/dataset-entries?where=... - filterable list of data set entry results
* /api/dataset-entries?where=....csv - filterable list of data set entry results, downloadable in csv format
* Some kind of Power BI link generator with appropriate content?

### Components

* `BaseLayout.vue` - see previous definition
* `DatasetLayout.vue` - see previous definition
* `DatasetVisualizeManagePage.vue`
    * `VisualizePropertiesFormCard.vue`
        * Allow Downloaded to CSV? - VisualizationControl#is_dowloadable_as_csv
        * Limit data view on data search - VisualizationControl#has_search_restrictions
        * Fields - VisualizationControl#has_search_field_restrictions
        * Exclude: - multi select of VisualizationControlSearchFieldExclusion#id
        * Rows - VisualizationControl#has_search_row_limits
        * (number input for row limit) - VisualizationControl#search_row_limit_maximum, has no visible label
    * `DatasetEntriesCard.vue` - card wrapper for easier reuse in [Dataset Visualize Manage Page](#heading=h.9jtkxxgvl1iu), could also use a “v-sheet”?
        * `DatasetEntriesTableSearch.vue` - undefined search feature for table
        * `DatasetEntriesDownloadCsvButton.vue` - triggers download from /datasets/:slug.csv
        * `DatasetEntriesOpenInPowerBIButton.vue` - triggers open link in new tab, or is a download with a special format handler, whichever is easiest. Requires investigation: e.g. build a power BI worksheet and download it, leaving opening the worksheet to the OS. _Leave this till last._
        * `DatasetEntriesTable.vue`
            * Headers from DatasetFields or DatasetSchemas
            * Data pulled from DatasetEntries

### Source Wireframe

![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/9c67a4d6-cbdc-4dd0-af2c-573ec182b885)

# Implementation

Re-imagine dataset integration with external dataset api. Now it takes you to a page with a lot of fancy custom logic.
Re-imagine dataset visualization search field exclusions.
Add ability to import a dataset from a URL, and preview the data in a table, as long as the dataset fields match the fields in the dataset.

# Screenshots

Re-imagined dataset integration create interface
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/92b94167-313c-4e12-b02c-68671871ea2e)

Re-imagined dataset integration update interface
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/0f899a6b-e144-44cc-b6b7-ebc1295d3e76)

Visualize manage page (note that search and download is being handled later.
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/00a46db5-d3b3-4466-8d7d-62d9eed3bd43)

# Testing Instructions

1. Run the test suite via `dev test` (or `dev test_api`)
2. Boot the app via `dev up`
3. Log in to the app at http://localhost:8080
4. Go to the manage page of a dataset, e.g. http://localhost:8080/datasets/employee-directory/description/manage
5. Click the "Add api link".
6. Fill in the url. e.g. https://api.gov.yk.ca/directory/employees, https://api.gov.yk.ca/directory/departments, or https://api.gov.yk.ca/directory/divisions
7. Check that creation fails unless you provide the correct header key and value.
8. Add those things, and the try to create again. It will take a while, because its fetching the dataset. If it fails try again, its probably the connection timing out or something.
9. Once the initial dataset has been created, you have the option to review it an update the parsing logic.
10. Parse out the appropriate sub-data using JMESPath.
11. Import the dataset and return the the previous tab.
12. Note that the "Add API link" button now says "manage api link".
13. Go to the Access tab, and check that the Data Access tab is now read-only and has a "password" field for the lisence key.
14. Go to the Fields tab and make sure that at least some of the fields match your dataset.
15. Go to the Visualize tab and check out the new UI.
16. Check that the table below the TODO has content in it that you can preview the dataset entries from the external api integration.
